File: frontend/packages/contracts/src/marketData.ts
import { z } from "zod";
import {
  ExchangeSchema,
  IsoDateSchema,
  SegmentSchema,
  TimeframeSchema,
  UtcDateTimeSchema,
} from "./common";
import { IndexNameSchema } from "./strategy";

/** Stored candles of one timeframe: first and last IST date (NOVA-097). */
export const InstrumentCoverageSchema = z.strictObject({
  timeframe: TimeframeSchema,
  from: IsoDateSchema,
  to: IsoDateSchema,
});
export type InstrumentCoverage = z.infer<typeof InstrumentCoverageSchema>;

/**
 * A stock with candles in any timeframe. Prices come from daily bars, or from intraday bars rolled up
 * per IST day when there are none; `dataFrom`/`dataTo` span every timeframe (NOVA-097).
 */
export const InstrumentSchema = z
  .strictObject({
    symbol: z.string().min(1),
    name: z.string().min(1),
    exchange: ExchangeSchema,
    segment: SegmentSchema,
    sector: z.string().min(1),
    indices: z.array(IndexNameSchema),
    lastClosePaise: z.number().int().positive(),
    high52wPaise: z.number().int().positive(),
    low52wPaise: z.number().int().positive(),
    changePercent: z.number(),
    avgDailyVolume: z.number().int().min(0),
    lotSize: z.number().int().positive().nullable(),
    timeframes: z.array(TimeframeSchema).min(1),
    dataFrom: IsoDateSchema,
    dataTo: IsoDateSchema,
    coverage: z.array(InstrumentCoverageSchema).min(1),
  })
  .refine((i) => i.coverage.map((c) => c.timeframe).join() === i.timeframes.join(), {
    message: "coverage must list the timeframes in the same order",
    path: ["coverage"],
  })
  .refine((i) => i.coverage.every((c) => c.from <= c.to), {
    message: "coverage from must be on or before to",
    path: ["coverage"],
  })
  .refine((i) => new Set(i.timeframes).size === i.timeframes.length, {
    message: "timeframes must be unique",
    path: ["timeframes"],
  })
  .refine((i) => i.dataFrom <= i.dataTo, {
    message: "dataFrom must be on or before dataTo",
    path: ["dataFrom"],
  })
  .refine((i) => new Set(i.indices).size === i.indices.length, {
    message: "indices must be unique",
    path: ["indices"],
  })
  .refine((i) => i.low52wPaise <= i.lastClosePaise && i.lastClosePaise <= i.high52wPaise, {
    message: "lastClosePaise must be between low52wPaise and high52wPaise",
    path: ["lastClosePaise"],
  });
export type Instrument = z.infer<typeof InstrumentSchema>;

const PricePaiseSchema = z.number().int().positive();

/** One OHLC bar. `time` is `YYYY-MM-DD` for `1d`, UTC ISO datetime for intraday (D17). */
export const CandleSchema = z
  .strictObject({
    time: z.union([IsoDateSchema, UtcDateTimeSchema]),
    openPaise: PricePaiseSchema,
    highPaise: PricePaiseSchema,
    lowPaise: PricePaiseSchema,
    closePaise: PricePaiseSchema,
    volume: z.number().int().min(0),
  })
  .refine((c) => c.highPaise >= Math.max(c.openPaise, c.closePaise), {
    message: "highPaise must be at least open and close",
    path: ["highPaise"],
  })
  .refine((c) => c.lowPaise <= Math.min(c.openPaise, c.closePaise), {
    message: "lowPaise must be at most open and close",
    path: ["lowPaise"],
  });
export type Candle = z.infer<typeof CandleSchema>;

/** An NSE index (D56): `members` stocks of the list belong to it. */
export const MarketIndexSchema = z.strictObject({
  name: IndexNameSchema,
  kiteSymbol: z.string().min(1),
  members: z.number().int().min(0),
  updatedAt: UtcDateTimeSchema.nullable(),
});
export type MarketIndex = z.infer<typeof MarketIndexSchema>;

File: backend/libs/nova_contracts/src/nova_contracts/market_data.py
"""Instrument and Candle: mirror `frontend/packages/contracts/src/marketData.ts`."""

from typing import Annotated, Self

from pydantic import Field, model_validator

from nova_contracts.common import Contract, Exchange, IsoDate, Segment, Timeframe, UtcDateTime

# Any `market_indices` row (D56); the API checks the name exists on write.
IndexName = Annotated[str, Field(min_length=1, max_length=40, pattern=r"^[A-Z0-9 &-]+$")]
PositivePaise = Annotated[int, Field(gt=0)]
NonEmpty = Annotated[str, Field(min_length=1)]


class Instrument(Contract):
    symbol: NonEmpty
    name: NonEmpty
    exchange: Exchange
    segment: Segment
    sector: NonEmpty
    indices: list[IndexName]
    last_close_paise: PositivePaise
    high52w_paise: PositivePaise = Field(alias="high52wPaise")
    low52w_paise: PositivePaise = Field(alias="low52wPaise")
    change_percent: float
    avg_daily_volume: Annotated[int, Field(ge=0)]
    lot_size: PositivePaise | None
    timeframes: Annotated[list[Timeframe], Field(min_length=1)]
    data_from: IsoDate
    data_to: IsoDate

    @model_validator(mode="after")
    def _rules(self) -> Self:
        if len(set(self.timeframes)) != len(self.timeframes):
            raise ValueError("timeframes must be unique")
        if self.data_from > self.data_to:
            raise ValueError("dataFrom must be on or before dataTo")
        if len(set(self.indices)) != len(self.indices):
            raise ValueError("indices must be unique")
        if not self.low52w_paise <= self.last_close_paise <= self.high52w_paise:
            raise ValueError("lastClosePaise must be between low52wPaise and high52wPaise")
        return self


class Candle(Contract):
    """`time` is a `YYYY-MM-DD` date for daily bars and a UTC datetime for intraday bars (D17)."""

    time: IsoDate | UtcDateTime
    open_paise: PositivePaise
    high_paise: PositivePaise
    low_paise: PositivePaise
    close_paise: PositivePaise
    volume: Annotated[int, Field(ge=0)]

    @model_validator(mode="after")
    def _ohlc(self) -> Self:
        if self.high_paise < max(self.open_paise, self.close_paise):
            raise ValueError("highPaise must be at least open and close")
        if self.low_paise > min(self.open_paise, self.close_paise):
            raise ValueError("lowPaise must be at most open and close")
        return self


class MarketIndex(Contract):
    """An NSE index (D56): `members` stocks of the list belong to it."""

    name: IndexName
    kite_symbol: NonEmpty
    members: Annotated[int, Field(ge=0)]
    updated_at: UtcDateTime | None
